# -*- coding: utf-8 -*-
# Django settings for DPNK project.

# Author: Petr Dlouh√Ω <petr.dlouhy@auto-mat.cz>
#
# Copyright (C) 2017 o.s. Auto*Mat
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from django.contrib.auth import logout

from .models import UserProfile


def create_userprofile(strategy, details, response, user=None, *args, **kwargs):
    """Update user details using data from provider."""
    if user and kwargs["is_new"]:
        UserProfile.objects.create(
            user=user,
            sex=response.get("gender", "unknown"),
        )


# https://stackoverflow.com/questions/13018147/authalreadyassociated-exception-in-django-social-auth


def social_user(backend, uid, user=None, *args, **kwargs):
    provider = backend.name
    social = backend.strategy.storage.user.get_social_auth(provider, uid)
    if social:
        if user and social.user != user:
            logout(backend.strategy.request)
        elif not user:
            user = social.user
    return {
        "social": social,
        "user": user,
        "is_new": user is None,
        "new_association": False,
    }
